### PR TITLE
Refactoring related to obsolete attachments directory

### PR DESCRIPTION
**Pull request by andr3ej**
* No longer create obsolete "attachments" directory
* Remove redundant NodeAttachmentService param
* Add type param to MockNetwork.Factory to eliminate casts
* Use null not -1 for unforced node ID
* Remove redundant createNode args

### CODE REVIEW

File: node/src/test/kotlin/net/corda/node/messaging/TwoPartyTradeFlowTests.kt [hunk 1]
```diff
@@ -289,11 +289,10 @@ class TwoPartyTradeFlowTests {
     // Creates a mock node with an overridden storage service that uses a RecordingMap, that lets us test the order
     // of gets and puts.
     private fun makeNodeWithTracking(
-            networkMapAddr: SingleMessageRecipient?,
-            name: X500Name,
-            overrideServices: Map<ServiceInfo, KeyPair>? = null): MockNetwork.MockNode {
```
Comment (andr3ej) [inline node/src/test/kotlin/net/corda/node/messaging/TwoPartyTradeFlowTests.kt]: wasn't used

File: samples/network-visualiser/src/main/kotlin/net/corda/netmap/simulation/Simulation.kt [hunk 2]
```diff
@@ -153,13 +153,10 @@ abstract class Simulation(val networkSendManuallyPumped: Boolean,
 
     val mockNet = MockNetwork(networkSendManuallyPumped, runAsync)
     // This one must come first.
-    val networkMap: SimulatedNode
-            = mockNet.createNode(null, nodeFactory = NetworkMapNodeFactory, advertisedServices = ServiceInfo(NetworkMapService.type)) as SimulatedNode
-    val notary: SimulatedNode
-            = mockNet.createNode(networkMap.network.myAddress, nodeFactory = NotaryNodeFactory, advertisedServices = ServiceInfo(SimpleNotaryService.type)) as SimulatedNode
-    val regulators: List<SimulatedNode> = listOf(mockNet.createNode(networkMap.network.myAddress, start = false, nodeFactory = RegulatorFactory) as SimulatedNode)
-    val ratesOracle: SimulatedNode
-            = mockNet.createNode(networkMap.network.myAddress, start = false, nodeFactory = RatesOracleFactory, advertisedServices = ServiceInfo(NodeInterestRates.Oracle.type)) as SimulatedNode
```
Comment (andr3ej) [inline samples/network-visualiser/src/main/kotlin/net/corda/netmap/simulation/Simulation.kt]: RIP casts

File: test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt [hunk 3]
```diff
@@ -67,31 +64,27 @@ class MockNetwork(private val networkSendManuallyPumped: Boolean = false,
...
-                  private val defaultFactory: Factory = MockNetwork.DefaultFactory,
+                  private val defaultFactory: Factory<*> = MockNetwork.DefaultFactory,
                   private val initialiseSerialization: Boolean = true) {
-    val nextNodeId
-        get() = _nextNodeId
-    private var _nextNodeId = 0
-    val filesystem: FileSystem = Jimfs.newFileSystem(unix())
-    private val busyLatch: ReusableLatch = ReusableLatch()
+    var nextNodeId = 0
+        private set
```
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt]: i should have done it this way in the first place

File: test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt [hunk 4]
```diff
@@ -267,43 +260,35 @@ class MockNetwork(private val networkSendManuallyPumped: Boolean = false,
...
-    /**
-     * Returns a node, optionally created by the passed factory method.
-     * @param overrideServices a set of service entries to use in place of the node's default service entries,
-     * for example where a node's service is part of a cluster.
      * @param entropyRoot the initial entropy value to use when generating keys. Defaults to an (insecure) random value,
      * but can be overridden to cause nodes to have stable or colliding identity/service keys.
      * @param configOverrides add/override behaviour of the [NodeConfiguration] mock object.
      */
-    fun createNode(networkMapAddress: SingleMessageRecipient? = null, forcedID: Int = -1, nodeFactory: Factory = defaultFactory,
+    fun createNode(networkMapAddress: SingleMessageRecipient? = null, forcedID: Int? = null,
```
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt]: null lets use use elvis below which is more idiomatic, and no performance reason to use a magic number here

File: test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt [hunk 5]
```diff
@@ -267,43 +260,35 @@ class MockNetwork(private val networkSendManuallyPumped: Boolean = false,
...
+    /** Like the other [createNode] but takes a [Factory] and propagates its [MockNode] subtype. */
+    fun <N : MockNode> createNode(networkMapAddress: SingleMessageRecipient? = null, forcedID: Int? = null, nodeFactory: Factory<N>,
+                   start: Boolean = true, legalName: X500Name? = null, overrideServices: Map<ServiceInfo, KeyPair>? = null,
+                   entropyRoot: BigInteger = BigInteger.valueOf(random63BitValue()),
+                   vararg advertisedServices: ServiceInfo,
+                   configOverrides: (NodeConfiguration) -> Any? = {}): N {
+        val id = forcedID ?: nextNodeId++
         val config = testNodeConfiguration(
-                baseDirectory = path,
+                baseDirectory = baseDirectory(id).createDirectories(),
```
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt]: idempotent in the forcedID case

File: test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt [hunk 6]
```diff
@@ -267,43 +260,35 @@ class MockNetwork(private val networkSendManuallyPumped: Boolean = false,
...
                 myLegalName = legalName ?: getTestX509Name("Mock Company $id")).also {
             whenever(it.dataSourceProperties).thenReturn(makeTestDataSourceProperties("node_${id}_net_$networkId"))
             configOverrides(it)
         }
         return nodeFactory.create(config, this, networkMapAddress, advertisedServices.toSet(), id, overrideServices, entropyRoot).apply {
             if (start) {
-                configuration.baseDirectory.createDirectories()
                 start()
-                if (threadPerNode && networkMapAddress != null) networkMapRegistrationFuture.getOrThrow()
+                if (threadPerNode && networkMapAddress != null) networkMapRegistrationFuture.getOrThrow() // TODO: Manually-started nodes won't do this.
```
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt]: fixing this now would risk breaking many tests, and i don't want to have to back out the other changes if that happens

File: test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt [hunk 7]
```diff
@@ -267,43 +260,35 @@ class MockNetwork(private val networkSendManuallyPumped: Boolean = false,
...
         val config = testNodeConfiguration(
-                baseDirectory = path,
+                baseDirectory = baseDirectory(id).createDirectories(),
                 myLegalName = legalName ?: getTestX509Name("Mock Company $id")).also {
             whenever(it.dataSourceProperties).thenReturn(makeTestDataSourceProperties("node_${id}_net_$networkId"))
             configOverrides(it)
         }
         return nodeFactory.create(config, this, networkMapAddress, advertisedServices.toSet(), id, overrideServices, entropyRoot).apply {
             if (start) {
-                configuration.baseDirectory.createDirectories()
```
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt]: redundant

File: test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt [hunk 8]
```diff
@@ -340,20 +325,19 @@ class MockNetwork(private val networkSendManuallyPumped: Boolean = false,
 
     /**
      * Sets up a network with the requested number of nodes (defaulting to two), with one or more service nodes that
-     * run a notary, network map, any oracles etc. Can't be combined with [createTwoNodes].
```
Comment (andr3ej) [inline test-utils/src/main/kotlin/net/corda/testing/node/MockNode.kt]: no longer such a method

**Comment by mikehearn** [comment]
Closing for now to free up TC capacity. Please re-open when it's ready to be assigned a reviewer.

**Pull request closed without merging**